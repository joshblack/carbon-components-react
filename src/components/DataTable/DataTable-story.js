import React, { Component } from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import classNames from 'classnames';
import PropTypes from 'prop-types';

import DataTable, {
  Container,
  Table,
  TableHead,
  TableHeader,
  TableBody,
  TableRow,
  TableCell,
} from '../DataTable';
import PaginationV2 from '../PaginationV2';
import Button from '../Button';
import OverflowMenu from '../OverflowMenu';
import OverflowMenuItem from '../OverflowMenuItem';

const paginationProps = {
  pageSizes: [10, 20, 30, 40, 50],
};

const rows = [[{ name: 'Load Bala' }]];

const initialRows = [
  {
    id: 'a',
    name: 'Load Balancer 3',
    protocol: 'HTTP',
    something: '80',
    rule: 'Round Robin',
    attached_groups: 'Kevins VM Groups',
    status: 'Active',
  },
  {
    id: 'b',
    name: 'Load Balancer 1',
    protocol: 'HTTP',
    something: '80',
    rule: 'Round Robin',
    attached_groups: 'Maureens VM Groups',
    status: 'Active',
  },
  {
    id: 'c',
    name: 'Load Balancer 2',
    protocol: 'HTTP',
    something: '80',
    rule: 'Round Robin',
    attached_groups: 'Andrews VM Groups',
    status: 'Active',
  },
];

const headers = [
  {
    key: 'name',
    header: 'Name',
    isSortable: true,
  },
  {
    key: 'protocol',
    header: 'Protocol',
    isSortable: true,
  },
  {
    key: 'something',
    header: 'Something',
    isSortable: false,
  },
  {
    key: 'rule',
    header: 'Rule',
    isSortable: false,
  },
  {
    key: 'attached_groups',
    header: 'Attached Groups',
    isSortable: true,
  },
  {
    key: 'status',
    header: 'Status',
    isSortable: true,
  },
];

storiesOf('DataTable', module).addWithInfo(
  'Data Table',
  `
    Data table
  `,
  () => (
    <DataTable
      rows={initialRows}
      headers={headers}
      render={({ rows, headers, getHeaderProps }) => (
        <Table>
          <TableHead>
            <TableRow>
              {headers.map(header => (
                <TableHeader {...getHeaderProps({ header })}>
                  {header.header}
                </TableHeader>
              ))}
            </TableRow>
          </TableHead>
          <TableBody>
            {rows.map(row => (
              <TableRow key={row.id}>
                {row.cells.map(cell => (
                  <TableCell key={cell.id}>{cell.value}</TableCell>
                ))}
              </TableRow>
            ))}
          </TableBody>
        </Table>
      )}
    />
  )
  // () => (
  // <DataTable
  // rows={initialRows}
  // headers={headers}
  // render={({
  // rows,
  // headers,

  // getHeaderProps,
  // getRowProps,
  // getCellProps,
  // }) => (
  // <Table>
  // <TableHead>
  // <TableRow>
  // {headers.map(({ header }) => (
  // <TableHeader {...getHeaderProps({ header })}>
  // {header}
  // </TableHeader>
  // ))}
  // </TableRow>
  // </TableHead>
  // <TableBody>
  // {rows.map((row, i) => (
  // <TableRow key={i}>
  // {Object.keys(row).map(key => (
  // <TableCell key={key}>{row[key]}</TableCell>
  // ))}
  // </TableRow>
  // ))}
  // </TableBody>
  // </Table>
  // )}
  // />
  // )
);
// .addWithInfo(
// 'Expandable table',
// `
// Expandable table
// `,
// () => <ExpandableDataTable />
// );
