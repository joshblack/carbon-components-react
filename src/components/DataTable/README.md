# `DataTable`

## Usage

### Default Table with Sorting

### Table with Selection

#### Integration with Batch Actions

### Table with Expansion

#### Expansion for all rows

#### Expansion for each row
