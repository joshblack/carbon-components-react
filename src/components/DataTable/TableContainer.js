import wrapComponent from '../../tools/wrapComponent';

const TableContainer = wrapComponent({
  name: 'TableContainer',
  className: ['bx--data-table-v2-container'],
  type: 'div',
});

export default TableContainer;
