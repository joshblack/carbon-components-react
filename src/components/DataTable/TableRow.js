import wrapComponent from '../../tools/wrapComponent';

const TableRow = wrapComponent({
  name: 'TableRow',
  type: 'tr',
});

export default TableRow;
