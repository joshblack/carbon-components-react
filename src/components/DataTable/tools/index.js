export const getCellId = (rowId, header) => `${rowId}:${header}`;

export const composeEventHandlers = fns => (event, ...args) => {
  for (let i = 0; i < fns.length; i++) {
    if (event.defaultPrevented) {
      break;
    }
    if (typeof fns[i] === 'function') {
      fns[i](event, ...args);
    }
  }
};
