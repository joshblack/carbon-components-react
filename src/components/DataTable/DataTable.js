import PropTypes from 'prop-types';
import React from 'react';
import { composeEventHandlers } from './tools';
import { denormalize, normalize } from './tools/format';
import {
  getSortState,
  initialSortState,
  defaultSortRows,
  sortStates,
} from './tools/sorting';

/**
 * Data Tables are used to represent a collection of resources, displaying a
 * subset of their fields in columns, or headers. We prioritize direct updates
 * to the state of what we're rendering, so internally we end up normalizing the
 * given data and then denormalizing it when rendering.
 *
 * As a result, each part of the DataTable is accessible through look-up by id,
 * and updating the state of the single entity will cascade updates to the
 * consumer.
 */
export default class DataTable extends React.Component {
  static propTypes = {
    /**
     * The `rows` prop is where you provide us with a list of all the rows that
     * you want to render in the table. The only hard requirement is that this
     * is an array of objects, and that each object has a unique `id` field
     * available on it.
     */
    rows: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.string.isRequired,
      })
    ),

    /**
     * The `headers` prop represents the order in which the headers should
     * appear in the table. You can also specify whether or not the table should
     * be sortable by this header with the `isSortable` field. If this value is
     * true, we'll provide the necessary props for sorting through the
     * `getHeaderProps` method available in the `render` method.
     */
    headers: PropTypes.arrayOf(
      PropTypes.shape({
        key: PropTypes.string.isRequired,
        header: PropTypes.string.isRequired,
        isSortable: PropTypes.bool,
      })
    ).isRequired,

    /**
     * Provide an `onChange` hook where you can subscribe to internal state
     * changes of the component.
     */
    onChange: PropTypes.func,

    /**
     * Optional hook to manually control sorting of the rows. Currently, this
     * method takes in the following:
     *
     * - rowIds, an array of string ids for each row
     * - cellsById, an object where each key is a cell id
     * - direction, one of 'NONE', 'ASC', or 'DESC'
     * - key, the key for the header to sort by
     * - locale, the current locale
     *
     * And is responsible for returning an array of sorted rowIds
     */
    sortRows: PropTypes.func,

    /**
     * Provide a string for the current locale
     */
    locale: PropTypes.string,
  };

  static defaultProps = {
    sortRows: defaultSortRows,
    locale: 'en',
  };

  constructor(props) {
    super(props);

    const { rowIds, rowsById, cellsById } = normalize(
      props.rows,
      props.headers
    );
    this.state = {
      rowIds,
      rowsById,
      cellsById,

      sortDirection: initialSortState,
      sortHeader: null,

      // Copy over rowIds so the reference doesn't mutate the stored
      // `initialRowOrder`
      initialRowOrder: rowIds.slice(),
    };
  }

  componentWillReceiveProps(nextProps) {
    const { rowIds, rowsById, cellsById } = normalize(
      nextProps.rows,
      nextProps.headers
    );
    this.setState({
      rowIds,
      rowsById,
      cellsById,
      sortDirection: initialSortState,
      sortHeader: null,
      initialRowOrder: rowIds.slice(),
    });
  }

  getHeaderProps = ({ header, onClick, ...rest }) => {
    const { key } = header;
    const { sortDirection, sortHeader } = this.state;
    return {
      ...rest,
      key,
      isSortHeader: sortHeader === key,
      sortDirection,
      // No need to include `onKeyDown` here since the underlying element
      // is a button
      onClick: composeEventHandlers([this.handleSortBy(key), onClick]),
    };
  };

  handleSortBy = key => () => {
    this.setState(state => {
      const {
        rowIds,
        cellsById,
        sortDirection,
        sortHeader,
        initialRowOrder,
      } = state;
      const { locale, sortRows } = this.props;
      const nextSortDirection = getSortState(key, sortHeader, sortDirection);
      const nextRowIds =
        nextSortDirection !== sortStates.NONE
          ? sortRows({
              rowIds: rowIds.slice(),
              cellsById,
              direction: sortDirection,
              key,
              locale,
            })
          : initialRowOrder;
      return {
        sortHeader: key,
        sortDirection: nextSortDirection,
        rowIds: nextRowIds,
      };
    });
  };

  getRowProps = ({ row }) => ({
    key: row.id,
  });

  getCellProps = ({ cell }) => ({
    key: cell.id,
  });

  render() {
    const { children, render } = this.props;
    const { rowIds, rowsById, cellsById } = this.state;
    const renderProps = {
      rows: denormalize(rowIds, rowsById, cellsById),
      headers: this.props.headers,
      getHeaderProps: this.getHeaderProps,
      getRowProps: this.getRowProps,
      getCellProps: this.getCellProps,
    };

    if (render !== undefined) {
      return render(renderProps);
    }

    if (children !== undefined) {
      return children(renderProps);
    }

    return null;
  }
}
