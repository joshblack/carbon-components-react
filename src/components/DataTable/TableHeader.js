import cx from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';
import Icon from '../Icon';
import { sortStates } from './tools/sorting';

const translationKeys = {
  'icon.description': 'Sort arrow',
};

const translateWithId = key => translationKeys[key];

const TableHeader = ({
  children,
  onClick,
  isSortHeader,
  sortDirection,
  translateWithId: t,
  ...rest
}) => {
  const className = cx({
    'bx--table-sort-v2': true,
    'bx--table-sort-v2--active':
      isSortHeader && sortDirection !== sortStates.NONE,
    'bx--table-sort-v2--ascending':
      isSortHeader && sortDirection === sortStates.ASC,
  });
  return (
    <th>
      <button className={className} onClick={onClick} {...rest}>
        <span className="bx--table-header-label">{children}</span>
        <Icon
          className="bx--table-sort-v2__icon"
          name="caret--down"
          description={t('icon.description')}
        />
      </button>
    </th>
  );
};

TableHeader.propTypes = {
  children: PropTypes.node,
  onClick: PropTypes.func.isRequired,
  isSortHeader: PropTypes.bool.isRequired,
  sortDirection: PropTypes.oneOf(['ASC', 'DESC', 'NONE']),
  translateWithId: PropTypes.func.isRequired,
};

TableHeader.defaultProps = {
  translateWithId,
};

TableHeader.translationKeys = Object.keys(translationKeys);

export default TableHeader;
