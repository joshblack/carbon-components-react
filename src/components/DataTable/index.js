import DataTable from './DataTable';
import TableContainer from './TableContainer';
import Table from './Table';
import TableBody from './TableBody';
import TableCell from './TableCell';
import TableHead from './TableHead';
import TableHeader from './TableHeader';
import TableRow from './TableRow';

DataTable.TableContainer = TableContainer;
DataTable.Table = Table;
DataTable.TableBody = TableBody;
DataTable.TableCell = TableCell;
DataTable.TableHead = TableHead;
DataTable.TableHeader = TableHeader;
DataTable.TableRow = TableRow;

export {
  TableContainer,
  Table,
  TableBody,
  TableCell,
  TableHead,
  TableHeader,
  TableRow,
};

export default DataTable;
