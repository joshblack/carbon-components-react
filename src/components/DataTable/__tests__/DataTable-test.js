import React from 'react';
import DataTable, {
  Container,
  Table,
  TableHead,
  TableHeader,
  TableBody,
  TableRow,
  TableCell,
} from '../';
import { mount } from 'enzyme';

const rows = [
  {
    id: 'a',
    name: 'Load Balancer 3',
    protocol: 'HTTP',
    something: '80',
    rule: 'Round Robin',
    attached_groups: 'Kevins VM Groups',
    status: 'Active',
  },
  {
    id: 'b',
    name: 'Load Balancer 1',
    protocol: 'HTTP',
    something: '80',
    rule: 'Round Robin',
    attached_groups: 'Maureens VM Groups',
    status: 'Active',
  },
  {
    id: 'c',
    name: 'Load Balancer 2',
    protocol: 'HTTP',
    something: '80',
    rule: 'Round Robin',
    attached_groups: 'Andrews VM Groups',
    status: 'Active',
  },
];

const headers = [
  {
    key: 'name',
    header: 'Name',
  },
  {
    key: 'protocol',
    header: 'Protocol',
  },
  {
    key: 'something',
    header: 'Something',
  },
  {
    key: 'rule',
    header: 'Rule',
  },
  {
    key: 'attached_groups',
    header: 'Attached Groups',
  },
  {
    key: 'status',
    header: 'Status',
  },
];

describe('DataTable', () => {
  let mockProps;

  beforeEach(() => {
    mockProps = {
      rows,
      headers,
      render: ({
        rows,
        headers,
        getHeaderProps,
        getRowProps,
        getCellProps,
      }) => (
        <Table>
          <TableHead>
            <TableRow>
              {headers.map(header => (
                <TableHeader {...getHeaderProps({ header })}>
                  {header.header}
                </TableHeader>
              ))}
            </TableRow>
          </TableHead>
          <TableBody>
            {rows.map(row => (
              <TableRow {...getRowProps({ row })}>
                {row.cells.map(cell => (
                  <TableCell {...getCellProps({ cell })}>
                    {cell.value}
                  </TableCell>
                ))}
              </TableRow>
            ))}
          </TableBody>
        </Table>
      ),
    };
  });

  it('should render', () => {
    const wrapper = mount(<DataTable {...mockProps} />);
    expect(wrapper).toMatchSnapshot();
  });
});
