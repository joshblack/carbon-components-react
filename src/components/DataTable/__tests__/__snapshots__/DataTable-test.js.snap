// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTable should render 1`] = `
<DataTable
  headers={
    Array [
      Object {
        "header": "Name",
        "key": "name",
      },
      Object {
        "header": "Protocol",
        "key": "protocol",
      },
      Object {
        "header": "Something",
        "key": "something",
      },
      Object {
        "header": "Rule",
        "key": "rule",
      },
      Object {
        "header": "Attached Groups",
        "key": "attached_groups",
      },
      Object {
        "header": "Status",
        "key": "status",
      },
    ]
  }
  locale="en"
  render={[Function]}
  rows={
    Array [
      Object {
        "attached_groups": "Kevins VM Groups",
        "id": "a",
        "name": "Load Balancer 3",
        "protocol": "HTTP",
        "rule": "Round Robin",
        "something": "80",
        "status": "Active",
      },
      Object {
        "attached_groups": "Maureens VM Groups",
        "id": "b",
        "name": "Load Balancer 1",
        "protocol": "HTTP",
        "rule": "Round Robin",
        "something": "80",
        "status": "Active",
      },
      Object {
        "attached_groups": "Andrews VM Groups",
        "id": "c",
        "name": "Load Balancer 2",
        "protocol": "HTTP",
        "rule": "Round Robin",
        "something": "80",
        "status": "Active",
      },
    ]
  }
  sortRows={[Function]}
>
  <Table>
    <table
      className="bx--data-table-v2 bx--data-table-v2--zebra"
    >
      <TableHead>
        <thead>
          <TableRow>
            <tr>
              <TableHeader
                isSortHeader={false}
                key="name"
                onClick={[Function]}
                sortDirection="NONE"
                translateWithId={[Function]}
              >
                <th>
                  <button
                    className="bx--table-sort-v2"
                    onClick={[Function]}
                  >
                    <span
                      className="bx--table-header-label"
                    >
                      Name
                    </span>
                    <Icon
                      className="bx--table-sort-v2__icon"
                      description="Sort arrow"
                      fillRule="evenodd"
                      name="caret--down"
                      role="img"
                    >
                      <svg
                        aria-label="Sort arrow"
                        className="bx--table-sort-v2__icon"
                        fillRule="evenodd"
                        height="5"
                        name="caret--down"
                        role="img"
                        viewBox="0 0 10 5"
                        width="10"
                      >
                        <title>
                          Sort arrow
                        </title>
                        <path
                          d="M10 0L5 5 0 0z"
                          key="key0"
                        />
                      </svg>
                    </Icon>
                  </button>
                </th>
              </TableHeader>
              <TableHeader
                isSortHeader={false}
                key="protocol"
                onClick={[Function]}
                sortDirection="NONE"
                translateWithId={[Function]}
              >
                <th>
                  <button
                    className="bx--table-sort-v2"
                    onClick={[Function]}
                  >
                    <span
                      className="bx--table-header-label"
                    >
                      Protocol
                    </span>
                    <Icon
                      className="bx--table-sort-v2__icon"
                      description="Sort arrow"
                      fillRule="evenodd"
                      name="caret--down"
                      role="img"
                    >
                      <svg
                        aria-label="Sort arrow"
                        className="bx--table-sort-v2__icon"
                        fillRule="evenodd"
                        height="5"
                        name="caret--down"
                        role="img"
                        viewBox="0 0 10 5"
                        width="10"
                      >
                        <title>
                          Sort arrow
                        </title>
                        <path
                          d="M10 0L5 5 0 0z"
                          key="key0"
                        />
                      </svg>
                    </Icon>
                  </button>
                </th>
              </TableHeader>
              <TableHeader
                isSortHeader={false}
                key="something"
                onClick={[Function]}
                sortDirection="NONE"
                translateWithId={[Function]}
              >
                <th>
                  <button
                    className="bx--table-sort-v2"
                    onClick={[Function]}
                  >
                    <span
                      className="bx--table-header-label"
                    >
                      Something
                    </span>
                    <Icon
                      className="bx--table-sort-v2__icon"
                      description="Sort arrow"
                      fillRule="evenodd"
                      name="caret--down"
                      role="img"
                    >
                      <svg
                        aria-label="Sort arrow"
                        className="bx--table-sort-v2__icon"
                        fillRule="evenodd"
                        height="5"
                        name="caret--down"
                        role="img"
                        viewBox="0 0 10 5"
                        width="10"
                      >
                        <title>
                          Sort arrow
                        </title>
                        <path
                          d="M10 0L5 5 0 0z"
                          key="key0"
                        />
                      </svg>
                    </Icon>
                  </button>
                </th>
              </TableHeader>
              <TableHeader
                isSortHeader={false}
                key="rule"
                onClick={[Function]}
                sortDirection="NONE"
                translateWithId={[Function]}
              >
                <th>
                  <button
                    className="bx--table-sort-v2"
                    onClick={[Function]}
                  >
                    <span
                      className="bx--table-header-label"
                    >
                      Rule
                    </span>
                    <Icon
                      className="bx--table-sort-v2__icon"
                      description="Sort arrow"
                      fillRule="evenodd"
                      name="caret--down"
                      role="img"
                    >
                      <svg
                        aria-label="Sort arrow"
                        className="bx--table-sort-v2__icon"
                        fillRule="evenodd"
                        height="5"
                        name="caret--down"
                        role="img"
                        viewBox="0 0 10 5"
                        width="10"
                      >
                        <title>
                          Sort arrow
                        </title>
                        <path
                          d="M10 0L5 5 0 0z"
                          key="key0"
                        />
                      </svg>
                    </Icon>
                  </button>
                </th>
              </TableHeader>
              <TableHeader
                isSortHeader={false}
                key="attached_groups"
                onClick={[Function]}
                sortDirection="NONE"
                translateWithId={[Function]}
              >
                <th>
                  <button
                    className="bx--table-sort-v2"
                    onClick={[Function]}
                  >
                    <span
                      className="bx--table-header-label"
                    >
                      Attached Groups
                    </span>
                    <Icon
                      className="bx--table-sort-v2__icon"
                      description="Sort arrow"
                      fillRule="evenodd"
                      name="caret--down"
                      role="img"
                    >
                      <svg
                        aria-label="Sort arrow"
                        className="bx--table-sort-v2__icon"
                        fillRule="evenodd"
                        height="5"
                        name="caret--down"
                        role="img"
                        viewBox="0 0 10 5"
                        width="10"
                      >
                        <title>
                          Sort arrow
                        </title>
                        <path
                          d="M10 0L5 5 0 0z"
                          key="key0"
                        />
                      </svg>
                    </Icon>
                  </button>
                </th>
              </TableHeader>
              <TableHeader
                isSortHeader={false}
                key="status"
                onClick={[Function]}
                sortDirection="NONE"
                translateWithId={[Function]}
              >
                <th>
                  <button
                    className="bx--table-sort-v2"
                    onClick={[Function]}
                  >
                    <span
                      className="bx--table-header-label"
                    >
                      Status
                    </span>
                    <Icon
                      className="bx--table-sort-v2__icon"
                      description="Sort arrow"
                      fillRule="evenodd"
                      name="caret--down"
                      role="img"
                    >
                      <svg
                        aria-label="Sort arrow"
                        className="bx--table-sort-v2__icon"
                        fillRule="evenodd"
                        height="5"
                        name="caret--down"
                        role="img"
                        viewBox="0 0 10 5"
                        width="10"
                      >
                        <title>
                          Sort arrow
                        </title>
                        <path
                          d="M10 0L5 5 0 0z"
                          key="key0"
                        />
                      </svg>
                    </Icon>
                  </button>
                </th>
              </TableHeader>
            </tr>
          </TableRow>
        </thead>
      </TableHead>
      <TableBody>
        <tbody>
          <TableRow
            key="a"
          >
            <tr>
              <TableCell
                key="a:name"
              >
                <td>
                  Load Balancer 3
                </td>
              </TableCell>
              <TableCell
                key="a:protocol"
              >
                <td>
                  HTTP
                </td>
              </TableCell>
              <TableCell
                key="a:something"
              >
                <td>
                  80
                </td>
              </TableCell>
              <TableCell
                key="a:rule"
              >
                <td>
                  Round Robin
                </td>
              </TableCell>
              <TableCell
                key="a:attached_groups"
              >
                <td>
                  Kevins VM Groups
                </td>
              </TableCell>
              <TableCell
                key="a:status"
              >
                <td>
                  Active
                </td>
              </TableCell>
            </tr>
          </TableRow>
          <TableRow
            key="b"
          >
            <tr>
              <TableCell
                key="b:name"
              >
                <td>
                  Load Balancer 1
                </td>
              </TableCell>
              <TableCell
                key="b:protocol"
              >
                <td>
                  HTTP
                </td>
              </TableCell>
              <TableCell
                key="b:something"
              >
                <td>
                  80
                </td>
              </TableCell>
              <TableCell
                key="b:rule"
              >
                <td>
                  Round Robin
                </td>
              </TableCell>
              <TableCell
                key="b:attached_groups"
              >
                <td>
                  Maureens VM Groups
                </td>
              </TableCell>
              <TableCell
                key="b:status"
              >
                <td>
                  Active
                </td>
              </TableCell>
            </tr>
          </TableRow>
          <TableRow
            key="c"
          >
            <tr>
              <TableCell
                key="c:name"
              >
                <td>
                  Load Balancer 2
                </td>
              </TableCell>
              <TableCell
                key="c:protocol"
              >
                <td>
                  HTTP
                </td>
              </TableCell>
              <TableCell
                key="c:something"
              >
                <td>
                  80
                </td>
              </TableCell>
              <TableCell
                key="c:rule"
              >
                <td>
                  Round Robin
                </td>
              </TableCell>
              <TableCell
                key="c:attached_groups"
              >
                <td>
                  Andrews VM Groups
                </td>
              </TableCell>
              <TableCell
                key="c:status"
              >
                <td>
                  Active
                </td>
              </TableCell>
            </tr>
          </TableRow>
        </tbody>
      </TableBody>
    </table>
  </Table>
</DataTable>
`;
